In [3]:
import pandas as pd
import numpy as np

# <b>NOTES<b>

## <b>Ideas:<b>
    -Merge Wave1, Wave2 and Wave3 in order to look at the same population throughout and check the differences
    -
    -
    -

    
    
    

In [75]:
#Convert from .dta to .csv
Wave2 = pd.read_stata("Wave2.dta")
# df.to_csv('Wave2.csv')
# df = pd.read_csv("Wave2.csv", index_col="pid", delimiter=",", low_memory=False)
# display(Wave2.head(5))
Wave2.info()

Wave2 = Wave2[Wave2['w2_nc_outcome'] == "Successfully Interviewed"]
# display(Wave2.head(5))
Wave2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7051 entries, 0 to 7050
Columns: 177 entries, pid to w2_nc_intlng3
dtypes: category(163), int32(1), object(13)
memory usage: 2.0+ MB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 5676 entries, 4 to 7050
Columns: 177 entries, pid to w2_nc_intlng3
dtypes: category(163), int32(1), object(13)
memory usage: 1.7+ MB


We have now narrrowed down our dataset to only the people who were successfully interviewed.

In [76]:
Wave1 = pd.read_stata("Wave1.dta")
# display(Wave1.head(5))
Wave1.info()

Wave1 = Wave1[Wave1['w1_nc_outcome'] == "Successfully Interviewed"]
# display(Wave1.head(5))
Wave1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7073 entries, 0 to 7072
Columns: 170 entries, pid to w1_nc_intlng11
dtypes: category(157), int32(1), object(12)
memory usage: 2.0+ MB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 7073 entries, 0 to 7072
Columns: 170 entries, pid to w1_nc_intlng11
dtypes: category(157), int32(1), object(12)
memory usage: 2.0+ MB


In [77]:
Wave3 = pd.read_stata("Wave3.dta")
# display(Wave3.head(5))
Wave3.info()

Wave3 = Wave3[Wave3['w3_nc_outcome'] == "Successfully Interviewed"]
# display(Wave3.head(5))
Wave3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8098 entries, 0 to 8097
Columns: 177 entries, pid to w3_nc_intlng3
dtypes: category(163), int32(1), object(13)
memory usage: 2.3+ MB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 6130 entries, 4 to 8097
Columns: 177 entries, pid to w3_nc_intlng3
dtypes: category(163), int32(1), object(13)
memory usage: 1.8+ MB


In [78]:
frames = [Wave1, Wave2, Wave3]
temp = pd.merge(Wave1, Wave2, how='inner', on='pid')
df = pd.merge(temp, Wave3, how='inner', on='pid')

display(df.head(5))
df.info()

pid             w1_nc_outcome w1_nc_intrv_c w1_nc_intrv_d w1_nc_intrv_m  \
0  301067  Successfully Interviewed           609          15th           May   
1  301079  Successfully Interviewed           609          22nd           May   
2  301082  Successfully Interviewed           624           4th          June   
3  301086  Successfully Interviewed           644           4th          June   
4  301088  Successfully Interviewed           630          27th           May   

  w1_nc_intrv_y w1_nc_duration w1_nc_dob_m w1_nc_dob_y w1_nc_gen  ...  \
0          2020           15.2        July      1971.0     Woman  ...   
1          2020          15.37    February      1970.0     Woman  ...   
2          2020          25.27        June      1989.0       Man  ...   
3          2020          17.18        June      1972.0     Woman  ...   
4          2020          21.13         May      1978.0       Man  ...   

  w3_nc_smkprice_ld w3_nc_smkbrand_ld w3_nc_smkbrand_ld_o w3_nc_smkcigday  \
0               NaN               NaN                                 NaN   
1               NaN               NaN                                 NaN   
2               NaN               NaN                                 NaN   
3               NaN               NaN                                 NaN   
4               NaN               NaN                                 NaN   

  w3_nc_smkbrand_after w3_nc_smkbrand_after_o w3_nc_intres w3_nc_intlng1  \
0                  NaN                            Friendly        Sepedi   
1                  NaN                            Friendly        Sepedi   
2                  NaN                            Friendly      IsiXhosa   
3                  NaN                            Friendly       IsiZulu   
4                  NaN                             Hostile      IsiXhosa   

  w3_nc_intlng2 w3_nc_intlng3  
0           NaN           NaN  
1       Sesotho           NaN  
2           NaN           NaN  
3           NaN           NaN  
4       English           NaN  

[5 rows x 522 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4508 entries, 0 to 4507
Columns: 522 entries, pid to w3_nc_intlng3
dtypes: category(483), int32(1), object(38)
memory usage: 3.9+ MB


In [81]:

def getDuplicateColumns(df): 
    duplicateColumnNames = set()
    for x in range(df.shape[1]): 
        col = df.iloc[:, x]
        for y in range(x + 1, df.shape[1]): 
            otherCol = df.iloc[:, y] 
            if col.equals(otherCol): 
                duplicateColumnNames.add(df.columns.values[y]) 
    return list(duplicateColumnNames) 
    
df = df.drop(columns = getDuplicateColumns(df)) 
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4508 entries, 0 to 4507
Columns: 521 entries, pid to w3_nc_intlng3
dtypes: category(483), int32(1), object(37)
memory usage: 3.9+ MB


The reason that only 1 column is deleted is because the column values are not consistent across the 3 waves as explained below.

In [82]:
df_dates = df[['w1_nc_dob_y', 'w2_nc_dob_y', 'w3_nc_dob_y', 'w1_nc_dob_m', 'w2_nc_dob_m', 'w3_nc_dob_m', 'w1_nc_gen', 'w2_nc_gen', 'w3_nc_gen']].copy()
df_dates

w1_nc_dob_y w2_nc_dob_y w3_nc_dob_y w1_nc_dob_m w2_nc_dob_m w3_nc_dob_m  \
0         1971.0      1971.0      1971.0        July        July        July   
1         1970.0      1970.0      1970.0    February    February    February   
2         1989.0      1989.0      1989.0        June        June        June   
3         1972.0      1972.0      1972.0        June        June        June   
4         1978.0      1978.0      1978.0         May         May         May   
...          ...         ...         ...         ...         ...         ...   
4503      1989.0      1990.0      1989.0      August      August      August   
4504      1978.0      1978.0      1978.0   September     October   September   
4505      1976.0      1976.0      1976.0     January     January     January   
4506      1996.0      1996.0      1996.0        July        July        July   
4507      1971.0      1971.0      1971.0     January     January     January   

     w1_nc_gen w2_nc_gen w3_nc_gen  
0        Woman     Woman     Woman  
1        Woman     Woman     Woman  
2          Man       Man       Man  
3        Woman     Woman     Woman  
4          Man       Man       Man  
...        ...       ...       ...  
4503     Woman     Woman     Woman  
4504       Man       Man       Man  
4505     Woman     Woman     Woman  
4506     Woman     Woman     Woman  
4507       Man       Man       Man  

[4508 rows x 9 columns]

There are several cases where the values are inconsistent across the different waves. I believe this is an input error since many of the entires will match. For example, the year of birth in record #4503 is (1989, 1990, 1989) yet the month of birth and gender is consistent (August, Woman). 
To clean this up, we should look at the triplet of columns (that we are interested in - birth date, gender, etc) and set all 3 values to whichever one appears most often.